### PR TITLE
Support small BLOCK_SIZE_M while fixing some bugs.

### PR DESCRIPTION
**Pull request by quintinwang5**
For 16 bits datatype with VNNI. This restriction is
image
So for 16x64, we split it to 2x16x32.

### CODE REVIEW

File: python/triton/language/semantic.py [hunk 1]
```diff
@@ -1367,9 +1367,6 @@ def _validate_dtype(dtype, allowed_types, operand_name):
     assert lhs_rank == rhs_rank == 2 or lhs_rank == rhs_rank == 3, f"Both inputs must be either 2D or 3D; (lhs: {lhs.shape} vs rhs: {rhs.shape})"
     assert lhs.shape[-1].value == rhs.shape[
         -2].value, f"First input shape ({lhs.shape}) and second input shape {rhs.shape} are not compatible for matmul (second index of first shape ({lhs.shape[-1].value}) must be equal to first index of second shape ({rhs.shape[-2].value})"
-    assert lhs.shape[-2].value >= 16 and lhs.shape[-1].value >= 16 \
```
Comment (etiotto) [inline python/triton/language/semantic.py:1370]: why is this assert removed ?

File: python/tutorials/09-experimental-block-pointer.py [hunk 2]
```diff
@@ -119,6 +119,9 @@
                       num_warps=32),
         triton.Config({'BLOCK_SIZE_M': 256, 'BLOCK_SIZE_N': 256, 'BLOCK_SIZE_K': 64, 'GROUP_SIZE_M': 4}, num_stages=4,
                       num_warps=32),
+        # For mxnxk=4x4096x512
```
Comment (etiotto) [inline python/tutorials/09-experimental-block-pointer.py:122]: What is the motivation to add a new block shape?

File: third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp [hunk 3]
```diff
@@ -168,10 +168,45 @@ class LoadStorePrefetchOpConversion
...
       auto load = rewriter.create<TritonGEN::Matrix2DBlockLoadOp>(
           loc, vectorType, base, surfaceW, surfaceH, surfaceP, offsetX, offsetY,
           dataSize, blockWidth, blockHeight, vBlks, false /* transpose*/, vnni);
-      rewriter.replaceOp(op, bitcast(load, resType));
+      Value res = load.getRes();
+      if (vnni) {
+        // TODO: Current max supported ld/st size is 512DWs = 64 x 16 x w (for
+        // 16bits dtype) so the tailHeight is less than(or eqaul to) 32, just
+        // need one more load
+        auto tailHeight = oldBlockHeight - maxLoadHeightVNNI;
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp:186]: auto -> unsigned ?

**Comment by Dewei-Wang-sh** [comment]
we'd better to add a lit test for the pass(convert-triton-to-tritongpu-warp, prefetch-block), as we are adding new supported shapes

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp [hunk 4]
```diff
@@ -254,8 +254,12 @@ createGenISA2DBlockRead(TritonGEN::Matrix2DBlockLoadOp op,
         op.getY(), i32_val(1));
     SmallVector<Type> argTypes{ptr_ty(context, 1), i32_ty, i32_ty, i32_ty,
                                vecType};
-    SmallVector<Value> args{op.getPtr(), op.getBaseWidth(), op.getBaseHeight(),
-                            op.getBasePitch(), byteCoord};
+    Value one = i32_val(1);
+    // We use minus 1 encoding in Gen lower pass,
+    // need to add it back for OpenCL call
+    SmallVector<Value> args{op.getPtr(), add(op.getBaseWidth(), one),
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:260]: As discussed, let's not add one in GEN to OCL lowering. But remove sub one from the user, and move that to GEN to GenISA lowering.

Comment (Dewei-Wang-sh) [reply third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:260]: when spirv built-ins are used in the future, are they adding one or not？

**Comment by quintinwang5** [comment]
Per @whitneywhtsang 's suggestions, I'll split this PR to 3 PRs.
1. Add `ceil` for `division` to avoid `0` in our `sizePerThread`/`ThreadsPerWarp`.
2. Push back `minus one` encoding to `GenISA` lowering. Not adding it back during `OpenCL` functions lowering
3. Other bugs fixing and small-k support.

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp (hunk 4 above)
Comment (whitneywhtsang) [reply third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:260]: It is not yet designed, but likely they are almost 1-1 mapping of the OCL built-ins.

File: third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp [hunk 5]
```diff
@@ -168,10 +168,45 @@ class LoadStorePrefetchOpConversion
...
+          std::iota(indices.begin(), indices.begin() + maxLoadHeightVNNI / 2,
+                    0);
+          std::iota(indices.begin() + maxLoadHeightVNNI / 2, indices.end(),
+                    maxLoadHeightVNNI);
+          auto attr = rewriter.getDenseI32ArrayAttr(indices);
+          res = rewriter.create<LLVM::ShuffleVectorOp>(
+              loc, load.getRes().getType(), load, tailLoad, attr);
+        }
+      }
+      rewriter.replaceOp(op, bitcast(res, resType));
```
Comment (Dewei-Wang-sh) [inline third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp:209]: if we need 2 loads, we should split it in match-target-size, the conversion pass serves for one op conversion.

File: python/triton/language/semantic.py (hunk 1 above)
Comment (quintinwang5) [reply python/triton/language/semantic.py:1370]: We support `M=8` now(or less in the future). But this assertion require `M/N/K >= 16`

**Comment by quintinwang5** [comment]
> Per @whitneywhtsang 's suggestions, I'll split this PR to 3 PRs.
> 
> 1. Add `ceil` for `division` to avoid `0` in our `sizePerThread`/`ThreadsPerWarp`.
> 2. Push back `minus one` encoding to `GenISA` lowering. Not adding it back during `OpenCL` functions lowering
> 3. Other bugs fixing and small-k support.

The three PRs are : [1132](https://github.com/intel/intel-xpu-backend-for-triton/pull/1132) [1137](https://github.com/intel/intel-xpu-backend-for-triton/pull/1137) [1152](https://github.com/intel/intel-xpu-backend-for-triton/pull/1152)
Close this PR as it's not needed now.

**Pull request closed without merging**